text/3838-repr-scalable.md

```diff
@@ -0,0 +1,386 @@
+and exists only enables scalable vector types to be defined in the standard
+library. The specific vector types are intended to eventually be stabilised, but
+none are proposed in this RFC.
+
+## Using scalable vectors
+[using-scalable-vectors]: #using-scalable-vectors
+
+Scalable vector types correspond to vector registers in hardware with unknown
+size at compile time. However, it will be a known and fixed size at runtime.
+Additional properties could be known during compilation, depending on the
+architecture, such as a minimum or maximum size or that the size must be a
+multiple of some factor.
+
+As previously described, users will not define their own scalable vector types
+and instead use intrinsics from `std::arch`, and this RFC is not proposing any
+such intrinsics, just the infrastructure. However, to illustrate how the types
+and intrinsics that this infrastructure will enable can be used, consider the
+following example that sums two input vectors:
+
+```rust
+fn sve_add(in_a: Vec<f32>, in_b: Vec<f32>, out_c: &mut Vec<f32>) {
+    let len = in_a.len();
+    unsafe {
+        // `svcntw` returns the actual number of elements that are in a 32-bit
+        // element vector
+        let step = svcntw() as usize;
+        for i in (0..len).step_by(step) {
+            let a = in_a.as_ptr().add(i);
+            let b = in_b.as_ptr().add(i);
+            let c = out_c as *mut f32;
+            let c = c.add(i);
+
+            // `svwhilelt_b32` generates a mask based on comparing the current
+            // index against the `len`
+            let pred = svwhilelt_b32(i as _, len as _);
+
+            // `svld1_f32` loads a vector register with the data from address
+            // `a`, zeroing any elements in the vector that are masked out
+            let sva = svld1_f32(pred, a);
+            let svb = svld1_f32(pred, b);
+
+            // `svadd_f32_m` adds `a` and `b`, any lanes that are masked out will
+            // take the keep value of `a`
+            let svc = svadd_f32_m(pred, sva, svb);
+
+            // `svst1_f32` will store the result without accessing any memory
+            // locations that are masked out
+            svst1_f32(svc, pred, c);
+        }
+    }
+}
+```
+
+From a user's perspective, writing code for scalable vectors isn't too different
+from when writing code with a fixed sized vector.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Types annotated with the `#[repr(simd)]` attribute contains either an array
+field or multiple fields to indicate the intended size of the SIMD vector that
+the type represents.
+
+Similarly, a `scalable(N)` representation is introduced to define a scalable
+vector type. `scalable(N)` accepts an integer to determine the minimum number of
+elements the vector contains. For example:
+
+```rust
+#[repr(simd, scalable(4))]
+pub struct svfloat32_t { _ty: [f32], }
```

Comment:
For example, this is misleading as this looks like it is unsized to begin with, but these are unconst `Sized`.

Reply:
I've changed the attribute to no longer be part of the `repr` attribute, and also changed the type of the type marker to no longer be an `!Sized` type, as these types will end up being non-const `Sized`.